cli: implement --trace-env and --trace-env-[js|native]-stack

joyeecheung opened the pull request:
This implements --trace-env, --trace-env-js-stack and --trace-env-native-stack CLI options which can be used to find out what environment variables are accessed and where they are accessed.

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

legendecas commented on src/node_credentials.cc:
```diff
@@ -87,14 +85,41 @@ bool SafeGetenv(const char* key,
+      fprintf(stderr, "undefined\n");
+    }
+
+    if (options->trace_env_native_stack) {
```
Can we use `PrintTraceEnvStack` here as well?

joyeecheung replied:
Done, used the EnvironmentOptions as argument instead since this place can also be hit when an Environment is not yet created (nullptr here).

joyeecheung commented:
Updated this a bit and added tracing for enumeration. I figured it would also be useful to see where e.g. the `...process.env` accesses are coming from.

targos commented:
Shouldn't collaborators be allowed to start CI without approving reviews?

legendecas commented:
> Shouldn't collaborators be allowed to start CI without approving reviews?

There was an update on the PR after the https://github.com/nodejs/node/labels/request-ci was added, which failed the check [here](https://github.com/nodejs/node-core-utils/blob/819e66955541afed455d6b067417c4651c700a5e/lib/pr_checker.js#L537-L540).

joyeecheung commented:
Fixed test to accomondate different builds.

joyeecheung commented:
Not sure why coverage-windows is failing but that seems to predate this PR: https://github.com/nodejs/node/actions/workflows/coverage-windows.yml

joyeecheung commented:
@legendecas can you take a look again? Thanks

legendecas commented on doc/api/cli.md:
```diff
@@ -2588,6 +2588,44 @@ added: v0.8.0
+* Queries in the form of `Object.hasOwn(process.env, 'KEY')`,
+  `process.env.hasOwnProperty('KEY')` or `'KEY' in process.env`.
+* Deletions in the form of `delete process.env.KEY`.
+* Enumerations inf the form of `...process.env` or `Object.keys(process.env)`.
```
This will potentially unvail credentials in the env (e.g. dotenv) once it has been accessed to the stdout. Would this be a security issue that is worth mentioning?

jasnell replied:
hmm, yeah this is potentially a bit problematic if folks don't expect it. Perhaps by default the `--trace-env*` flags shouldn't actually include the values (or should replace the values with something like `****`, and have the user explicitly use an option like `--trace-env=values` to explicitly opt-in to printing the values for each key.

joyeecheung replied:
Updated to just not print the values at all - at least for the use cases intended so far (checking consistency), this does not seem to be particularly important. We can add another flag or something to print the value when someone requests it.

KSDLightermm commented:
- [ ]

lavang1412 commented:
joyeecheung:trace-env
